tools/src/main/java/org/nmdp/ngs/tools/ValidateInterpretation.java

```diff
@@ -47,105 +49,118 @@
-            Map<String, SubjectTyping> observed = readObserved(observedFile,lociList);	
-   
-            for (String sample : expected.keySet()) {
-            	
-            	if(observed.get(sample) != null){        		
-            		
-            		for(String loc : lociList){
-
-		                List<String> alleles = expected.get(sample).getTyping(loc);
-		                List<String> interpretations = observed.get(sample).getTyping(loc);
-	
-		                if(interpretations.size() != 0){
-			                for (String expectedAllele : alleles) {
-			                    boolean result = false;
-			                    
-			                    for (String interpretation : interpretations) {
-		                  	
-			                        List<String> interpretedAlleles = Splitter.onPattern("[/|]+").splitToList(interpretation);
-			                        List<String> found = new ArrayList<String>();
-			                                     
-			                        for (String interpretedAllele : interpretedAlleles) {
-			                            if (matchByField(expectedAllele, interpretedAllele) >= resolution) {
-			                                found.add(interpretedAllele);
-			                            }                            
-			                        }
-			
-			                        if (!found.isEmpty()) {
-			                            result = true;
-			                        }
-			                    }
-			                    
-	
-			                    if (result) {
-			                        passes++;
-			                    }
-			                    else {
-			                        failures++;
-			                    }
-			
-			                    if (!printSummary) {
-			                        writer.println((result ? "PASS" : "FAIL") + "\t" + sample + "\t" + expectedAllele + "\t" + expected.get(sample).getExperiment());
-			                    }
-			                }
-	            		}
-            		}
-            	}
-            }
 
+            // interpretations keyed by sample
+            ListMultimap<String, Interpretation> expected = read(expectedFile);
+            ListMultimap<String, Interpretation> observed = read(observedFile);
+
+            // for each sample in observed
+            for (String sample : observed.keySet()) {
+                for (Interpretation o : observed.get(sample)) {
+                    AlleleList observedAlleleList = asAlleleList(o);
+                    if (shouldValidate(o, observedAlleleList)) {
+
+                        // for each matching sample in expected
+                        for (Interpretation e : expected.get(sample)) {
+                            Genotype expectedGenotype = asGenotype(e);
+                            for (Haplotype expectedHaplotype : expectedGenotype.getHaplotypes()) {
+                                for (AlleleList expectedAlleleList : expectedHaplotype.getAlleleLists()) {
```

Comment:
If there is ambiguity in the expected glstring then there will be a validated result for each allele in the expected glstring, right? This is fine for now since we're not dealing with any ambiguity in the expected glstring. Once allele code expansion is added to this then we'll only want one validated result for the whole expected glstring. Meaning it would be a pass if every allele in the expected is found in the observed.